clients/src/test/java/org/apache/kafka/clients/consumer/internals/CoordinatorTest.java

```diff
@@ -247,69 +257,132 @@ public void testCoordinatorDisconnect() {
     }
 
     @Test
-    public void testNormalJoinGroup() {
-        subscriptions.subscribe(Arrays.asList(topicName), subscriptionListener);
+    public void testNormalJoinGroupLeader() {
+        final String consumerId = "leader";
+
+        subscriptions.subscribe(Arrays.asList(topicName), rebalanceListener);
         subscriptions.needReassignment();
 
+        // ensure metadata is up-to-date for leader
+        metadata.setTopics(Arrays.asList(topicName));
+        metadata.update(cluster, time.milliseconds());
+
         client.prepareResponse(consumerMetadataResponse(node, Errors.NONE.code()));
         coordinator.ensureCoordinatorKnown();
 
         // normal join group
-        client.prepareResponse(joinGroupResponse(1, "consumer", Collections.singletonList(tp), Errors.NONE.code()));
+        Map<String, List<String>> memberSubscriptions = Collections.singletonMap(consumerId, Arrays.asList(topicName));
+        partitionAssignor.prepare(Collections.singletonMap(consumerId, Arrays.asList(tp)));
+
+        client.prepareResponse(joinGroupLeaderResponse(1, consumerId, memberSubscriptions, Errors.NONE.code()));
+        client.prepareResponse(new MockClient.RequestMatcher() {
+            @Override
+            public boolean matches(ClientRequest request) {
+                SyncGroupRequest sync = new SyncGroupRequest(request.request().body());
+                return sync.memberId().equals(consumerId) &&
+                        sync.generationId() == 1 &&
+                        sync.groupAssignment().containsKey(consumerId);
+            }
+        }, syncGroupResponse(Arrays.asList(tp), Errors.NONE.code()));
         coordinator.ensurePartitionAssignment();
 
         assertFalse(subscriptions.partitionAssignmentNeeded());
         assertEquals(Collections.singleton(tp), subscriptions.assignedPartitions());
-        assertEquals(1, subscriptionListener.revokedCount);
-        assertEquals(Collections.emptySet(), subscriptionListener.revoked);
-        assertEquals(1, subscriptionListener.assignedCount);
-        assertEquals(Collections.singleton(tp), subscriptionListener.assigned);
+        assertEquals(1, rebalanceListener.revokedCount);
+        assertEquals(1, rebalanceListener.assignedCount);
+        assertEquals(Collections.singleton(tp), rebalanceListener.assigned);
```

Comment:
Woops, I'll add the assertion back. I think at one point I had added code to only invoke the revocation callback when we were leaving a valid generation, so the expected revokeCount was actually zero and the check on revoked was unneeded. Later on, I removed the check to be consistent with the current code, but forgot this assertion.